src/main/java/io/odpf/firehose/sink/bigquery/proto/ProtoUpdateListener.java

```diff
@@ -0,0 +1,104 @@
+import io.odpf.firehose.sink.bigquery.converter.MessageRecordConverter;
+import io.odpf.firehose.sink.bigquery.converter.MessageRecordConverterWrapper;
+import io.odpf.firehose.sink.bigquery.converter.RowMapper;
+import io.odpf.firehose.sink.bigquery.handler.BigQueryClient;
+import io.odpf.firehose.sink.bigquery.models.BQField;
+import io.odpf.firehose.sink.bigquery.models.ProtoField;
+import lombok.Setter;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.stream.Collectors;
+
+@Slf4j
+public class ProtoUpdateListener extends com.gojek.de.stencil.cache.ProtoUpdateListener {
+    private final BigQuerySinkConfig config;
+    private final Converter protoMappingConverter = new Converter();
+    private final ProtoFieldParser protoMappingParser = new ProtoFieldParser();
+    private final BigQueryClient bqClient;
+    private final MessageRecordConverterWrapper messageRecordConverterWrapper;
+    @Setter
+    private Parser stencilParser;
+
+    public ProtoUpdateListener(BigQuerySinkConfig config, BigQueryClient bqClient, MessageRecordConverterWrapper messageRecordConverterWrapper) {
+        super(config.getInputSchemaProtoClass());
+        this.config = config;
+        this.bqClient = bqClient;
+        this.messageRecordConverterWrapper = messageRecordConverterWrapper;
+    }
+
+    public void update(Map<String, DescriptorAndTypeName> newDescriptors) {
+        onProtoUpdate("", newDescriptors);
+    }
+
+    @Override
+    public void onProtoUpdate(String url, Map<String, DescriptorAndTypeName> newDescriptors) {
+        log.info("stencil cache was refreshed, validating if bigquery schema changed");
+        try {
+            ProtoField protoField = new ProtoField();
+            protoField = protoMappingParser.parseFields(protoField, config.getInputSchemaProtoClass(), StencilUtils.getAllProtobufDescriptors(newDescriptors), StencilUtils.getTypeNameToPackageNameMap(newDescriptors));
+            updateProtoParser(protoField);
+        } catch (BigQueryException | IOException e) {
+            String errMsg = "Error while updating bigquery table on callback:" + e.getMessage();
+            log.error(errMsg);
+            throw new RuntimeException(errMsg, e);
+        }
+    }
+
+    // First get latest protomapping, update bq schema, and if all goes fine
+    // then only update beast's proto mapping config
+    private void updateProtoParser(final ProtoField protoField) throws IOException {
+        String protoMappingString = protoMappingConverter.generateColumnMappings(protoField.getFields());
+        List<Field> bqSchemaFields = protoMappingConverter.generateBigquerySchema(protoField);
+        addMetadataFields(bqSchemaFields);
+        bqClient.upsertTable(bqSchemaFields);
+        config.setProperty("SINK_BIGQUERY_PROTO_COLUMN_MAPPING", protoMappingString);
```

Comment:
Strange, seems this proto column mapping is not being used anywhere else other than passed as argument,  and also this is the only place we need to change a config value using `setProperty`
i wonder whether we need to set this config or not, 
or even further do we even need to make the config mutable ?

Reply:
basically, there is a converter which converts string to properties.